engine/lib/filestore.php

```diff
@@ -598,7 +576,206 @@ function elgg_get_inline_url(\ElggFile $file, $use_cookie = false, $expires = fa
+		],
+		'medium' => [
+			'w' => 153,
+			'h' => 153,
+			'square' => true,
+			'metadata_name' => 'smallthumb',
+			'filename_prefix' => 'smallthumb',
+		],
+		'large' => [
+			'w' => 600,
+			'h' => 600,
+			'square' => false,
+			'metadata_name' => 'largethumb',
+			'filename_prefix' => 'largethumb',
+		],
+	];
+
+	// @todo: add a hook to filter these values
+	return $sizes;
+}
+
+/**
+ * Create thumbnails for an image file
+ *
+ * @param \ElggFile $file File entity
+ * @return bool
+ */
+function elgg_create_thumbnails(\ElggFile $file) {
+
+	$prefix = 'file/';
+	$filestorename = elgg_substr($file->getFilename(), elgg_strlen($prefix));
+
+	$sizes = elgg_get_thumbnail_sizes($file);
+
+	foreach ($sizes as $size => $opts) {
+		$metadata_name = $opts['metadata_name'];
+		$filename_prefix = $opts['filename_prefix'];
+		$width = $opts['w'];
+		$height = $opts['h'];
+		$square = $opts['square'];
+		if (!isset($file->$metadata_name)) {
+			$file->$metadata_name = $prefix . $filename_prefix . $filestorename;
+		}
+		$thumb = elgg_get_thumbnail($file, $size);
+		$thumbnail = get_resized_image_from_existing_file($file->getFilenameOnFilestore(), $width, $height, $square);
+		if ($thumbnail) {
+			$thumb->open("write");
+			$thumb->write($thumbnail);
+			$thumb->close();
+			unset($thumbnail);
+		} else {
+			elgg_clear_thumbnails($file);
+			return false;
+		}
+	}
+
+	$file->icontime = time();
+	return true;
+
+}
+
+/**
+ * Clear file entity's thumbnails
+ *
+ * @param \ElggFile $file File entity
+ * @return void
+ */
+function elgg_clear_thumbnails(\ElggFile $file) {
+
+	$sizes = elgg_get_thumbnail_sizes($file);
+	foreach ($sizes as $size => $opts) {
+		$thumb = elgg_get_thumbnail($file, $size);
+		$thumb->delete();
+		$metadata_name = $opts['metadata_name'];
+		unset($file->$metadata_name);
+	}
+
+	unset($file->icontime);
+}
+
+/**
+ * Returns a file that represents file entity's thumbnail on filestore
+ * Note that the thumbnail file may or may not exist, use ``ElggFile::exists()``
+ * on this function's return to ensure that the file has an actual thumbnail on filestore
+ * 
+ * @param \ElggFile $file File entity (must have a guid)
+ * @param string    $size Thumnail size
+ * @return \ElggFile
+ * @since 2.2
+ */
+function elgg_get_thumbnail(\ElggFile $file, $size = 'medium') {
+
+	$sizes = elgg_get_thumbnail_sizes($file);
+	if (!array_key_exists($size, $sizes)) {
+		$size = 'large';
+	}
+
+	$metadata_name = $sizes[$size]['metadata_name'];
+
+	$thumbnail = new \ElggFile();
+	$thumbnail->owner_guid = $file->owner_guid;
+	$thumbnail->setFilename($file->$metadata_name);
+
+	return $thumbnail;
+}
 
-return function(\Elgg\EventsService $events, \Elgg\HooksRegistrationService $hooks) {
+return function(\Elgg\EventsService $events, \Elgg\HooksRegistrationService $hooks, \Elgg\Router $router) {
 	$events->registerHandler('init', 'system', '_elgg_filestore_init', 100);
+	$router->registerPageHandler('embed-file', '_elgg_filestore_embed_file_handler');
```

Comment:
Why can't we register this in `_elgg_filestore_init`? We wouldn't need to add the router.